### PR TITLE
[BEAM-10402] Enable checkerframework globally

### PR DESCRIPTION
**Pull request by kennknowles**
Instead of opting out whole modules now only existing classes are opted out of type checking. This has the following benefits:

 - New code will be checked, even in modules that are not yet passing.
 - Code won't as easily backslide in modules that were partially fixed.
 - It will produce much more manageable incremental work, and limitless starter bugs!

I produced the needed warnings (merged in other PRs) by removing the flag and then repeating the following, more or less:

```
./gradlew compileJava compileTestJava --continue 2>&1 | tee ~/tmp/gradle.log
 
cat ~/tmp/gradle.log \
    | grep -e "$PWD"'.*error:.*' \
    | cut -d : -f 1 \
    | sort -u \
    | xargs grep -L -e '^@SuppressWarnings' \
    | xargs sed -E -i '' 's/^(public +)?((abstract|final) +)?(class|interface)/@SuppressWarnings("nullness") &/'
  
./gradlew spotlessApply
```

There are two ways to still suppress type checking, arguments to `applyJavaNature`:

1. `generatedClassPatterns` to exclude various generated code that is not annotated with nullness types
2. `classesTriggerCheckerBugs` a map from classes which cannot be analyzed to their checkerframework bug URL

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by kennknowles** [comment]
@TheNeuralBit I don't know why I think you would enjoy / be amused by this, but I chose arbitrarily. @jayendra13 may also be interested.

```
% find . -name '*.java' -or -name 'build.gradle' | wc -l
    5424
```

So this involved suppression or tweaking on about 2k out of 5.5k files. You can image how it is important to get this done before we grow even bigger...

**Comment by kennknowles** [comment]
Oh now I remember - there was some runtime ramification of enabling checker for certain modules. Do you recall?

**Comment by jayendra13** [comment]
I am definitely interested in this, but got stuck with my own work. Will resume soon may be in a week or two 🙂 .

**Comment by kennknowles** [comment]
New files since I ran it, of course. It may not be possible to review without that happening during review. It doesn't mean much for the review of the concept, I think.

**Comment by kennknowles** [comment]
The NPE reproduced on `InMemoryJobServiceTest` when mockito tried to mock it. I believe it is a bug in the JDK. Annotations that are not present on the classpath should be "as if they were not there" but instead the JDK NPEs. Maybe they should use checker to eliminate those null errors :-)

Since the annotations are MIT licensed they can be `compile` scope instead of `compileOnly` scope. That fixed `InMemoryJobServiceTest`.

File: examples/java/src/test/java/org/apache/beam/examples/complete/TrafficRoutesIT.java [hunk 1]
```diff
@@ -42,6 +42,7 @@
 
 /** End-to-end tests of TrafficRoutes. */
 @RunWith(JUnit4.class)
+@SuppressWarnings("nullness")
```
Comment (TheNeuralBit) [inline examples/java/src/test/java/org/apache/beam/examples/complete/TrafficRoutesIT.java:45]: Could we add a TODO(BEAM-10402) on these?

**Review by TheNeuralBit — COMMENTED** [review]
The concept looks good. Do you think you could change it so all of the @SuppressWarnings annotations are added in a separate commit? Then we'd just be left with the gradle changes, and any other java changes that make sense to review manually.

**Comment by kennknowles** [comment]
Great ideas.

I learned a bit from:

 - https://stackoverflow.com/questions/20929176/how-can-i-find-modified-lines-that-match-a-certain-pattern-with-git
 - https://stackoverflow.com/questions/36932309/how-do-i-git-add-only-lines-matching-a-pattern 
 - https://stackoverflow.com/questions/13955390/git-grep-but-restricted-to-new-or-modified-files-in-the-index
 - https://stackoverflow.com/questions/1751397/git-grep-cached

In particular, I did not know about `git grep` or `git diff-index` at all. Now I do, I guess? I never did figure out exactly how to do what you suggest in one go. Perhaps the best way to check that I got it is to separate PRs...

The enabling (somewhat of a leap of faith?) which I will later `rebase -i` to come after the rest.

```
git reset HEAD^
git status | grep build.gradle | cut -d : -f 2 | xargs git add
git commit -m 'Enable checkerframework globally'
```

I figured out I could commit everything that wasn't a new suppression (interactively, somewhat tedious but far less than 2k files)

```
git add -p $(git diff-index -G ' .*Nullable.*' HEAD | cut -d ' ' -f 5 | cut -f 2)
```

And the API surface checks all include a new line of `pruningPattern`

```
git diff-index -SpruningPattern HEAD
```

(needed to tweak interactively)

```
git commit -m 'Allow checkerframework on API surfaces'
```

Confirmation that the only things remaining are suppressions at the top level:

```
git diff -U0 | grep '^[+][^+]' | grep -v '@SuppressWarnings'
```

(oops there were a few more, which I isolated and committed)

**Comment by kennknowles** [comment]
Gya, I botched the API surface patches versus moving nullable annotation patches

**Comment by kennknowles** [comment]
OK. The commits look right to me. Let me know if it helps to split things any other way. I had hoped this would be a pretty silly PR but it got slightly involved. I would say two things to check are:

1. all except the monster commit make sense
2. the monster commit contains nothing other than suppressions

Yea?

**Comment by kennknowles** [comment]
Run Java PreCommit

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 2]
```diff
@@ -1646,8 +1656,10 @@ class BeamModulePlugin implements Plugin<Project> {
       project.ext.applyJavaNature(
           exportJavadoc: false,
           enableSpotbugs: false,
-          enableChecker: false,
           publish: configuration.publish,
+          generatedClassPatterns: [
+            '^org\\.apache\\.beam\\..*'
+          ],
```
Comment (TheNeuralBit) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:1662]: This looked concerning until I looked around and realized the "portability" nature means its for making modules that generate protobuf/grpc code. The comment above "applyPortabilityNature should only be applied to projects that want to use vendored gRPC / protobuf dependencies" isn't very helpful. Maybe I'll send you a PR to document these natures.

**Comment by kennknowles** [comment]
Hmm this might slow the build down enough that it has to be separated from the CI build.

**Comment by kennknowles** [comment]
Run Python_PVR_Flink PreCommit

**Comment by kennknowles** [comment]
Run Python_PVR_Flink PreCommit

**Comment by kennknowles** [comment]
Run Java_Examples_Dataflow PreCommit

File: sdks/java/extensions/protobuf/build.gradle [hunk 3]
```diff
@@ -18,8 +18,13 @@
 
 plugins { id 'org.apache.beam.module' }
 applyJavaNature(
-  enableChecker: false,
   ignoreRawtypeErrors: true,
+  generatedClassPatterns: [
+    '^org\\.apache\\.beam\\.sdk\\.extensions\\.protobuf\\.Proto2CoderTestMessages',
+    '^org\\.apache\\.beam\\.sdk\\.extensions\\.protobuf\\.Proto2SchemaMessages',
+    '^org\\.apache\\.beam\\.sdk\\.extensions\\.protobuf\\.Proto3SchemaMessages',
+    '^org\\.apache\\.beam\\.sdk\\.extensions\\.protobuf\\.Proto3SchemaOptions',
```
Comment (TheNeuralBit) [inline sdks/java/extensions/protobuf/build.gradle:26]: Is it possible to use [slashy strings](https://groovy-lang.org/syntax.html#_slashy_string) in our gradle files? That'd clean these arrays up a bit.

File: sdks/java/extensions/sql/hcatalog/build.gradle [hunk 4]
```diff
@@ -18,7 +18,12 @@
 
 plugins { id 'org.apache.beam.module' }
 
-applyJavaNature(enableChecker:false,automaticModuleName: 'org.apache.beam.sdk.extensions.sql.meta.provider.hcatalog')
+applyJavaNature(
+  automaticModuleName: 'org.apache.beam.sdk.extensions.sql.meta.provider.hcatalog',
+  classesTriggerCheckerBugs: [
+    'HCatalogTable': '',
```
Comment (TheNeuralBit) [inline sdks/java/extensions/sql/hcatalog/build.gradle:24]: not yet filed or TODO

File: sdks/java/io/hadoop-format/build.gradle [hunk 5]
```diff
@@ -20,7 +20,7 @@ import groovy.json.JsonOutput
 
 plugins { id 'org.apache.beam.module' }
 applyJavaNature(
-  enableChecker:false,
+
```
Comment (TheNeuralBit) [inline sdks/java/io/hadoop-format/build.gradle:23]: nit: some of these are left with unnecessary whitespace (not a big deal if its a pain to fix)

File: sdks/java/io/kudu/build.gradle [hunk 6]
```diff
@@ -18,8 +18,12 @@
 
 plugins { id 'org.apache.beam.module' }
 applyJavaNature(
-  enableChecker: false,
+
   ignoreRawtypeErrors:true,
+  classesTriggerCheckerBugs: [
+    'KuduTestUtils': '',
+    'KuduIOIT': '',
```
Comment (TheNeuralBit) [inline sdks/java/io/kudu/build.gradle:25]: not yet filed or TODO

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 7]
```diff
@@ -726,11 +729,15 @@ class BeamModulePlugin implements Plugin<Project> {
...
+        options.compilerArgs -= [
+          "-Xlint:deprecation",
+        ]
         options.compilerArgs += ([
           '-parameters',
           '-Xlint:all',
-          '-Werror'
+          '-Werror',
+          '-Xmaxerrs',
+          '10000'
```
Comment (TheNeuralBit) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:740]: Is this a typo for maxerrors?

I'm assuming you added this to make it quicker to run, can/should we drop it now?

**Review by TheNeuralBit — APPROVED** [review]
LGTM, thank you! This will be much better, hopefully seeing the SuppressWarnings annotations will inspire some more crowdsourced fixes.

I did a quick check over the monster commit to make sure it was just SuppressWarnings, and looked over the other one more closely.

I'm assuming you'll have to re-generate the monster commit to sync with master, can you add TODOs at the same time?

**Comment by kennknowles** [comment]
Seeing really wonky errors on Jenkins that I cannot repro locally. Could be checkerframework bugs related to particular JDKs?

**Comment by kennknowles** [comment]
Run Java PreCommit

**Comment by kennknowles** [comment]
Run Java PreCommit

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 2 above)
Comment (kennknowles) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Yea. I could be `.model` except it is also used by the windmill proto generation. I have changed it so that it is passed in and the default is the model.

File: examples/java/build.gradle [hunk 8]
```diff
@@ -21,7 +21,7 @@ import groovy.json.JsonOutput
 plugins { id 'org.apache.beam.module' }
 applyJavaNature(
   exportJavadoc: false,
-  enableChecker: false,
+
```
Comment (kennknowles) [inline examples/java/build.gradle:24]: Noting that we don't apply spotless to the groovy files but we can and it should clean up these blanks and the indentation inconsistencies etc.

File: sdks/java/extensions/protobuf/build.gradle (hunk 3 above)
Comment (kennknowles) [reply sdks/java/extensions/protobuf/build.gradle]: Didn't know about slashy strings. Nice. Done.

File: sdks/java/extensions/sql/hcatalog/build.gradle (hunk 4 above)
Comment (kennknowles) [reply sdks/java/extensions/sql/hcatalog/build.gradle]: Done

File: sdks/java/io/kudu/build.gradle (hunk 6 above)
Comment (kennknowles) [reply sdks/java/io/kudu/build.gradle]: Done

**Review by kennknowles — COMMENTED** [review]
Added the TODOs. Slashy strings work but I discovered that the eclipse groovy formatter used by spotless does not understand them. That is why I did not use them in `BeamModulePlugin`.

File: sdks/java/extensions/protobuf/build.gradle (hunk 3 above)
Comment (TheNeuralBit) [reply sdks/java/extensions/protobuf/build.gradle]: Nice! I was unaware of them before, but did a check to see if there was some kind of regex literal we could use. I guess `~"pattern"` is an option as well, but produces `Pattern` instances.

**Comment by kennknowles** [comment]
I'm not saying there isn't a problem with this PR, but this setup is strange. The command `./gradlew pullLicenses` essentially is doing a full build as part of that single task, and the two builds do not coordinate so they can clobber/corrupt files. Suspect that this PR just changes the timings so that it happens reliably. Filed https://issues.apache.org/jira/browse/BEAM-11116 and I will see if I can resolve it. So far, the commands run by the script don't seem to work as expected anyhow...

**Comment by kennknowles** [comment]
In #13189 I develop a fix and simplification. I see that with and without that fix, `pullLicenses` does not seem to re-run all the targets. So I would conclude that something in this PR causes the subprocess call to build a bunch more stuff. I don't know what it is.

**Comment by kennknowles** [comment]
Since there are lingering issues, one way to make this PR more manageable is to break out the monster commit and get it in: #13191.

**Comment by kennknowles** [comment]
Confirmed with `git bisect` that this change causes the java compile.

```diff
--- a/sdks/java/container/build.gradle
+++ b/sdks/java/container/build.gradle
@@ -82,7 +82,7 @@ licenseReport {
               project.rootProject.findProject(':sdks:java:io:kafka')]
   excludeOwnGroup = true
   excludeGroups = ["beam.*"] // project dependencies do not match their maven coords
-  configurations = ALL
+  configurations = ["runtime", "shadow", "compile"]
   renderers = [new JsonReportRenderer()]
 }
```

I did not also bisect but I also guess that `excludes = ['org.checkerframwork:jdk8']` causes the compile. That line doesn't work anyhow, because it excludes deps too late.

**Comment by kennknowles** [comment]
Excluding the dep from all configurations is safe and effective, not requiring any alteration to the docker container license pull logic.

**Comment by kennknowles** [comment]
The changes to the license logic are separated in #13189 but I added the commits here to get precommit signal while waiting on review there.

**Comment by kennknowles** [comment]
Run Portable_Python PreCommit

**Comment by kennknowles** [comment]
Java precommit seems to have degraded in runtime.

 - Before: https://scans.gradle.com/s/gohimd5ekkcmo/timeline
 - After: https://scans.gradle.com/s/z4brjsxfpftxy/timeline

So it went from 40 minutes to 90 minutes. It isn't apples-to-apples as we have a lot of caching effects since they are just running on Jenkins:

 - Before: `2375 tasks executed in 116 projects in 34m 18.746s, with 162 avoided tasks saving 1h 19.326s`
 - After: `2463 tasks executed in 116 projects in 1h 20m 11.508s, with 38 avoided tasks saving 14m 48.735s`

I think zero-cache builds in some semi-controlled environment might give a better sense whether this needs to be treated specially.

**Comment by kennknowles** [comment]
I'm doing something like that on my laptop. Certainly the core SDK compile is slower, but even with a huge slowdown in compile the test execution should dominate. So I do think it may be caching that is the difference. Running it a few times to warm the build cache might also give a sense of realistic CI times.

**Comment by kennknowles** [comment]
run java precommit

**Comment by kennknowles** [comment]
Run Java PreCommit

**Comment by kennknowles** [comment]
Yea the re-run took 30 minutes so this might be a non-problem.

**Comment by kennknowles** [comment]
https://scans.gradle.com/s/smgegyyrcnoua/timeline but I'm not sure I believe the statement "2463 tasks executed in 116 projects in 26m 4.055s, with 174 avoided tasks *saving 5h 43m 25.832s*"

**Comment by kennknowles** [comment]
run java precommit

**Comment by TheNeuralBit** [comment]
Is it all the build.gradle changes invalidating the cache?

**Comment by kennknowles** [comment]
TBH not sure

**Comment by kennknowles** [comment]
Yea on the bad run it is claiming that compiling the SDK took 16 minutes and compiling the SDK's tests took 30 minutes. So that is orders of magnitude more than what I can repro. Something is awry. I'm going to try a rebase which should un-cache those and we can see if it was isolated. I suspect it was not.

**Comment by kennknowles** [comment]
The re-run was 38 minutes, pretty normal. Many things pulled from cache. https://scans.gradle.com/s/wyh5rgkbd3rjq/timeline?sort=longest

"2463 tasks executed in 116 projects in 33m 22.938s, with 173 avoided tasks saving 4h 12m 58.349s"

**Comment by kennknowles** [comment]
Here are some samples of runs against commits and phrases:

 - https://scans.gradle.com/s/acfajsh5mnxj2/timeline?sort=longest
 - https://scans.gradle.com/s/wbqztstpu7pdo/timeline?sort=longest
 - https://scans.gradle.com/s/7qqcuygtqlaou/timeline?sort=longest

It is hard to determine the real likely slowdown in practice. The tasks that were the longest and most problematic in the bad run are almost always `FROM-CACHE`.

**Comment by kennknowles** [comment]
On my idle desktop machine:

 - `git checkout github/master && ./gradlew :javaPreCommit --continue --no-build-cache --rerun-tasks --scan` https://scans.gradle.com/s/cdfrii6qsopii/timeline?sort=longest
 - `git checkout checkerframework-classlevel && ./gradlew :javaPreCommit --continue --no-build-cache --rerun-tasks --scan` https://scans.gradle.com/s/2r65rr5mflsro/timeline?sort=longest

Unfortunate that the build seems very unhealthy on master so I am just looking at particular tasks. It does look like compile tasks are perhaps 5x as slow but it does not affect the overall runtime very much. One option is to cease analyzing test java. These seem to have a much greater slowdown factor. I wonder if incorrect types are more expensive than correct ones...

**Comment by jayendra13** [comment]
We have fixed some of the module, so I think creating a map somewhere denoting modules fixed and not fixed would help for better tracking this issue. What do you think? I am soon going to start this again.

**Comment by TheNeuralBit** [comment]
> I think creating a map somewhere denoting modules fixed and not fixed would help for better tracking this issue

Are you thinking something like a spreadsheet that we would keep up-to-date, or would this be a part of the build.

With the change in https://github.com/apache/beam/pull/13191 we could identify fixed vs. not fixed classes and/or modules by grepping for `@SuppressWarnings("nullness") // TODO(https://issues.apache.org/jira/browse/BEAM-10402)`. Maybe that's enough for tracking? We could put a bash one-liner in the BEAM-10402 comments

**Comment by jayendra13** [comment]
> Are you thinking something like a spreadsheet that we would keep up-to-date, or would this be a part of the build?
Not part of a build, but some where publicly accessible like kanban board on jira or on github .

**Comment by kennknowles** [comment]
OK this is finally healthy and ran in a normal amount of time (precommits tend to be in the 30-60 minute range and this was 30). Only the tests for SDK core are skipped. Hopefully we won't have to skip them forever, since carelessness with nulls in tests is an annoying source of false issues. Going to merge it!

**Comment by kennknowles** [comment]
Confirmed that there are no breakages introduced in master since the tests ran here.

**Comment by kennknowles** [comment]
Since there are 2000+ classes with issues, probably this is too many Jiras and too much for a kanban? I don't know. Actually each class can be enough of a challenge for a Jira. But maybe Jiras for the module level is easier to manage.

A spreadsheet containing the output of a bash script that finds the `TODO` might make a lot of sense for tracking the burndown.
